Feature/aapd 617 sql schema

beejjacobs opened the pull request:
## Ticket Number

<!-- Add the number from the Jira board -->

https://pins-ds.atlassian.net/browse/AAPD-617

## Description of change

Add an initial SQL schema with `AppealUser` as a focus. Updated the seed script too.

## Checklist

<!-- Put an `x` in all the boxes that apply: -->

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [ ] My commit history in this PR is linear
- [ ] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

beejjacobs commented on packages/appeals-service-api/.env.example:
```diff
@@ -4,4 +4,4 @@
 
-DATABASE_URL="sqlserver://0.0.0.0:1433;database=pins_front_office_development;user=sa;password=DockerDatabaseP@22word!;trustServerCertificate=true"
\ No newline at end of file
+DATABASE_URL="sqlserver://localhost:1433;database=pins_front_office_development;user=sa;password=DockerDatabaseP@22word!;trustServerCertificate=true"
```
I needed this for it to work for me.

bryn-pins replied:
Should have caught this in last PR, this isn't the naming I've used for Azure
Can change here or there
Maybe why dev is down?

https://github.com/Planning-Inspectorate/infrastructure-environments/blob/3134210d37501df66e0b20d4ef64704854093db8/app/components/appeals-app-services/locals.tf#L94

beejjacobs replied:
Good catch - no the current issue on dev - but it will be!

We specify the env var to use in the schema file (datasource.url). I don't mind much which we use, but we need to align.

beejjacobs replied:
Going to update this in this PR to align with infra, which has a more specific name.

PaulGruffyddAmaze commented on packages/appeals-service-api/src/db/schema.prisma:
```diff
@@ -10,8 +7,119 @@ datasource db {
+  // link to an AppealUser, who is the agent
+  // this is optional because not all appeals have an agent
+  agentUserId String?     @db.UniqueIdentifier
+  Agent       AppealUser? @relation("agent", fields: [agentUserId], references: [id], onDelete: NoAction, onUpdate: NoAction)
```
According to the PINS data model, `agent` is an array rather than a single entity so I wonder whether we should create a cross-reference table to allow the linking of multiple agents to an appeal. 

If we were to also add a field to the cross reference table for `role`, that would allow us to remove references to all types of users from the appeal table and manage the relationships between entities independently. That would also allow us to manage other types of relation between those entities e.g. `interestedParty`. Obviously the trade-off there is we'd need a table join to retrieve the data rather than querying a single table.

beejjacobs replied:
Great point! I think a cross reference table is the right solution. I've implemented that.

PaulGruffyddAmaze commented on packages/appeals-service-api/src/db/schema.prisma:
```diff
@@ -10,8 +7,120 @@ datasource db {
+// This is a many-to-many relationship.
+model AppealToUser {
+  // relation to an Appeal
+  appealId   String     @unique @db.UniqueIdentifier
```
I don't think `appealId` would be unique in this instance as we can have multiple users (e.g. agent & appellant) linked to the same appeal.

beejjacobs replied:
Definitely a copy-paste error!

PaulGruffyddAmaze commented on packages/appeals-service-api/src/db/schema.prisma:
```diff
@@ -10,8 +7,120 @@ datasource db {
+  appealId   String     @unique @db.UniqueIdentifier
+  Appeal     Appeal     @relation(fields: [appealId], references: [id])
+  // relation to a User
+  userId     String     @unique @db.UniqueIdentifier
```
Similar to my comment above, a given user (particularly an agent) may be associated with multiple appeals so I don't think `userId` would be unique in this table.

beejjacobs replied:
Definitely a copy-paste error!

bryn-pins commented on packages/appeals-service-api/src/db/schema.prisma:
```diff
@@ -10,8 +7,120 @@ datasource db {
+  id         String  @id @default(dbgenerated("newid()")) @db.UniqueIdentifier
+  email      String  @unique
+  // role may only be needed for lpaUser, and may not be populated otherwise
+  role       String? // appellant | agent | lpaUser
```
Thinking again I guess this is correct and we do want a role on both AppealUser and AppealToUser as lpa users are not linked to appeals directly

However I guess this will just be lpauser | null 
at least to start with as an appeal user could be linked as either an appellant or an agent at the same time

beejjacobs replied:
Replaced in favour of `isLpaUser`

bryn-pins commented on packages/appeals-service-api/src/db/schema.prisma:
```diff
@@ -10,8 +7,120 @@ datasource db {
+  AppealUser AppealUser @relation(fields: [userId], references: [id])
+
+  // the role the user has in relation to the appeal
+  role String // appellant | agent | interestedParty
```
potentially use this a foreign key to a role table for clarity?

beejjacobs replied:
Done.

bryn-pins commented on packages/appeals-service-api/src/db/schema.prisma:
```diff
@@ -10,8 +7,120 @@ datasource db {
+
+// AppealToUser represents the relationships between AppealUser and Appeal
+// This is a many-to-many relationship.
+model AppealToUser {
```
I feel as if maybe AppealUser should just be called User?
That way this table name makes more sense (should be AppealUserToAppeal at the moment)
And LPA users aren't directly linked to an Appeal
Not that important though

beejjacobs commented:
Made some changes:

* Added `AppealToUserRole` table as an enum for user role
* Added `seed-dev` and `seed-prod` scripts, with some reference data
* use `SQL_CONNECTION_STRING` to match infra

bryn-pins reviewed (commented):
minor comment

beejjacobs commented on packages/appeals-service-api/src/db/seed/data-dev.js:
```diff
@@ -0,0 +1,36 @@
+/**
+ * @type {import('@prisma/client').Prisma.AppealUserCreateInput}
+ */
+const users = [
```
d'oh - yes.

beejjacobs replied:
fixed
